lib/parsers/yarn-utils.ts

```diff
@@ -0,0 +1,72 @@
+import {structUtils} from '@yarnpkg/core';
+import * as _ from 'lodash';
+
+const BUILTIN_PLACEHOLDER = 'builtin';
+const MULTIPLE_KEYS_REGEXP = / *, */g;
+
+export type ParseDescriptor = typeof structUtils.parseDescriptor;
+export type ParseRange = typeof structUtils.parseRange;
+
+const keyNormalizer = (parseDescriptor: ParseDescriptor, parseRange: ParseRange) =>
+    (rawDescriptor: string): string[] => {
+        // See https://yarnpkg.com/features/protocols
+        const descriptors: string[] = [rawDescriptor];
+        const descriptor = parseDescriptor(rawDescriptor);
+        const name = `${descriptor.scope ? '@' + descriptor.scope + '/' : ''}${descriptor.name}`;
+        const range = parseRange(descriptor.range);
+        const protocol = range.protocol;
+        switch (protocol) {
+            case 'npm:':
+            case 'file:':
+                descriptors.push(`${name}@${range.selector}`);
+                descriptors.push(`${name}@${protocol}${range.selector}`);
+                break;
+            case 'http:':
+            case 'https:':
+            case 'link:':
+            case 'portal:':
+            case 'exec:':
+            case 'workspace:':
+            case 'virtual:':
+                descriptors.push(`${name}@${protocol}${range.selector}`);
+                break;
```

Comment:
An alternative would be to put this in the `default` block - probably more future proof, since users can define their own protocols through the plugin API.

Reply:
Yes it thought about user defined plugins, so you suggest to not throw in case of unknown protocol at all?